File: implementations/dotnet/EmailInquire/Validators/CommonProviderMistake.cs
using System.Linq;

namespace EmailInquire.Validators
{
    /// <summary>
    ///     Validation with hint result over Damerau-Levenshtein algorithm and common domains list
    /// </summary>
    public class CommonProviderMistake : ValidatorBase<CommonProviderMistake>
    {
        public override Response Validate()
        {
            if (CommonDomains.Value.Contains(Domain)) return Response.Undefined;
            var match = CommonDomains.Value.FirstOrDefault(d => Distance(d, Domain) == 1);
            return match != null ? Response.Hint(Name, match) : Response.Undefined;
        }
    }
}

File: implementations/ruby/lib/email_inquire/validator/common_provider_mistake.rb
# frozen_string_literal: true

require "damerau-levenshtein"
require "email_inquire/validator/base"
require "email_inquire/validator/common_provider"

module EmailInquire
  module Validator
    class CommonProviderMistake < Base

      def validate
        return if CommonProvider::DOMAINS.include?(domain)

        replacement_domain =
          CommonProvider::DOMAINS.find do |reference|
            distance = DamerauLevenshtein.distance(domain, reference)

            distance.equal?(1)
          end

        response.hint!(domain: replacement_domain) if replacement_domain
      end

    end
  end
end
